_posts/2021-09-06-terraform-uncommon-options.markdown

```diff
@@ -0,0 +1,336 @@
+
+```hcl
+locals {
+  map = {
+    a = 1
+    b = 2
+    c = 3
+  }
+}
+
+resource "local_file" "private_key" {
+  content  = jsonencode(local.map)
+  filename = "./map.json"
+}
+```
+Needless to say, executing `terraform init` is necessary to pull the plugins and initialize the configured backend if there is any.
+
+Afterwads running `terraform apply` against this template, a JSON encoded file with the content of the local variable `map` will be created in the same folder. That is easy enough.
+
+But what if I needed a YAML encoded file? I am also not sure how it will look alike. I prefer JSON, but I also like to have some comments.
+
+I can check out on the fly by firing up Terraform console in the same folder and just YAML encoding the local `map` variable:
+
+```sh
+$ terraform console
+> yamlencode(local.map)
+<<EOT
+"a": 1
+"b": 2
+"c": 3
+
+EOT
+```
+Voila! We see how our file will look like if we've used the `yamlencode` function instead.
+
+Why stop there, maybe I just need the value of the `b` key?
+
+```sh
+> yamlencode(local.map.b)
+<<EOT
+2
+...
+
+EOT
+```
+
+Swapping the keys and the values is also something that I might want?
+
+```sh
+> yamlencode({for k,v in local.map : v => k })
+<<EOT
+"1": "a"
+"2": "b"
+"3": "c"
+
+EOT
+```
+Or just "trying" out the `try` function to see how it exactly works.
+
+```sh
+> try(yamlencode(local.map.d), 4)
+4
+```
+
+I digress. The flexibility that the `terraform console` provides when testing, troubleshooting, or simply playing around is immense.
+
+## [Shell Autocomplete]
+
+One of the commands that do not appear when invoking terraform via the CLI. Myself having worked with Terraform for quite a while, I stumbled upon this command fairly recently. IT lets you use shell tab-completion. As of writing this blog post, it works with `bash` or `zsh` shells only.
+
+Running this will enable autocomplete for terraform CLI. Re-reading the shell configuration or re-reloading a shell will activate it.
+```sh
+$ terraform -install-autocomplete
+```
+
+## [Lock Timeout]
+
+It is another very useful option, especially when running in an automated environment, such as a CI system. Imagine the scenario: we start two or more plan or apply jobs at a similar time. Naturally, terraform will fail if it cannot get a lock for the state. With the option `-lock-timeout` instead of immediately failing, Terraform will retry getting the state lock for the duration specified in seconds. If Terraform cannot acquire a lock after the timeout has passed, it will fail with an error.
+
+```sh
+$ terraform apply/plan -lock-timeout=180s
+```
+
+## [Manual State Manipulation]
+Usually, I edit the state via Terraform CLI. Sometimes, though, it is required that I need to manually edit it. I do it via a pull, edit, and push cycle. These commands will work regardless of the type of backend configured.
```

Comment:
To make your second sentence flow better, I'd change it to "Sometimes, though, I need to manually edit it."